src/fragment-generation-utils.js

```diff
@@ -437,16 +437,22 @@ const FragmentFactory = class {
 
     if (canExpandRange) {
       if (this.startOffset < this.getStartSearchSpace().length) {
-        // Find the next boundary char.
-        // TODO: should keep going if we haven't added any non boundary chars.
-        const newStartOffset = this.getStartSearchSpace()
-                                   .substring(this.startOffset + 1)
-                                   .search(fragments.internal.BOUNDARY_CHARS);
-        if (newStartOffset === -1) {
-          this.startOffset = this.getStartSearchSpace().length;
-        } else {
-          this.startOffset = this.startOffset + 1 + newStartOffset;
-        }
+        // Shift to the next boundary char, and repeat until we've added a word
+        // char.
+        const oldStartOffset = this.startOffset;
+        do {
```

Comment:
It looks like for each iteration, we do 4 traversals of the string: 2 substring operations and 2 search operations.

Would it be more efficient to simply traverse the searchSpace char by char until we reach the first non-boundary? Not sure if that would still be compatible with the algorithm. In any case it's not a huge deal because of the eventual shift to segmenter.

Reply:
In practice the substring and search operations in the `while` check should be fast because they're only on the new part we've added on this iteration.

We could probably switch this to some kind of state machine where we step through the characters, check each against the appropriate regexes, and halt when we find a word character after a boundary character. But I think the performance gains would be modest, and the focus right now is on the new implementation, so I'll keep that in mind as something if we see timeout stats diverging between the old and new implementations.